ipykernel/eventloops.py

```diff
@@ -302,12 +302,22 @@ def kernel_handler():
         loop.call_later(kernel._poll_interval, kernel_handler)
 
     loop.call_soon(kernel_handler)
-    try:
-        if not loop.is_running():
+    # loop is already running (e.g. tornado 5), nothing left to do
+    if loop.is_running():
+        return
+    while True:
+        error = None
+        try:
             loop.run_forever()
-    finally:
+        except KeyboardInterrupt:
+            continue
+        except Exception as e:
+            error = e
         loop.run_until_complete(loop.shutdown_asyncgens())
```

Comment:
The docs say that `shutdown_asyncgens` is new in 3.6, so this code might not work with 3.5? But maybe it's only on the abstract class that the method is new. This is already in the code, so I'm happy to merge this PR anyway, but it's worth checking.